### PR TITLE
optimize performance for MetricsFilter

### PR DESCRIPTION
**Pull request by finefuture**
ReflectionUtils add event type cache, reduce the time consumption of MetricsFilter

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Review by songxiaosheng — APPROVED** [review]
LGTM @wxbty  PATL

**Review by wxbty — COMMENTED** [review]
Good job,  write a unit test by the way？

**Comment by finefuture** [comment]
Why did it affect the dubbo-samples-triple-reactor module？
image

**Review by wxbty — APPROVED** [review]
LGTM

**Review by AlbumenJ — CHANGES_REQUESTED** [review]
image

How about cache in AbstractMetricsListener.

Allocate a map to cache class is not a good idea, and this may cause classloader unable to being gc.

**Comment by finefuture** [comment]
> image
> 
> How about cache in AbstractMetricsListener.
> 
> Allocate a map to cache class is not a good idea, and this may cause classloader unable to being gc.

@AlbumenJ Good idea, Is this ok?
image

**Comment by wxbty** [comment]
> > image
> > How about cache in AbstractMetricsListener.
> > Allocate a map to cache class is not a good idea, and this may cause classloader unable to being gc.
> 
> @AlbumenJ Good idea, Is this ok? image
> 
> Tested to improve performance by 2% than cache in ReflectionUtils

Would it be better to put the match method in util, and use Object for the event type? It is not the responsibility of the current class

**Comment by finefuture** [comment]
> > > image
> > > How about cache in AbstractMetricsListener.
> > > Allocate a map to cache class is not a good idea, and this may cause classloader unable to being gc.
> > 
> > 
> > @AlbumenJ Good idea, Is this ok? image
> > Tested to improve performance by 2% than cache in ReflectionUtils
> 
> Would it be better to put the match method in util, and use Object for the event type? It is not the responsibility of the current class

Okay, Moved

**Review by wxbty — APPROVED** [review]

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/AbstractMetricsListener.java [hunk 1]
```diff
@@ -20,16 +20,19 @@
 import org.apache.dubbo.common.utils.ReflectionUtils;
 import org.apache.dubbo.metrics.event.MetricsEvent;
 
-import java.util.List;
+import java.util.HashMap;
+import java.util.Map;
 
 public abstract class AbstractMetricsListener<E extends MetricsEvent> implements MetricsListener<E> {
 
+    private final Map<Class<?>, Boolean> eventMatchCache = new HashMap<>();
```
Comment (AlbumenJ) [inline dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/AbstractMetricsListener.java:28]: ConcurrentHashMap here

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/AbstractMetricsListener.java (hunk 1 above)
Comment (finefuture) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/AbstractMetricsListener.java]: Fixed

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**